probml_utils/logistic_regression.py

```diff
@@ -0,0 +1,137 @@
+import jax
+import jax.numpy as jnp 
+import optax
+
+@jax.jit
+def binary_loss_function(parameters, x, y, lambd):
+    '''
+    Arguments:
+        parameters: a dictionary contains weights and bias
+        x = datasets, shape (no. of examples, no. of features)
+        y = datasets, shape (no. of examples, )
+        lambd = regularization rate
+    
+    return:
+        loss - binary cross entropy loss
+    '''
+    z = jnp.dot(x,parameters["weights"]) + parameters["bias"]
+    hypothesis_x = jax.nn.sigmoid(z)
+    regularizer = (lambd * jnp.sum(jnp.sum(parameters["weights"]**2)))/(2*x.shape[0])
+
+    return (-((jnp.dot(y.T, jnp.log(hypothesis_x + 1e-7)) + jnp.dot((1-y).T ,jnp.log(1- hypothesis_x + 1e-7)))/x.shape[0]) + regularizer)[0]
+
+@jax.jit
+def multi_loss_function(parameters, x, y, lambd):
+    """
+    args:
+        parameters: a dictionary contains weights and bias
+        x = datasets, shape (no. of examples, no. of features)
+        y = datasets, shape (no. of examples, no. of classes)
+        lambd = regularization rate
+    
+    return:
+        loss = cross-entropy loss
+    """
+    z = jnp.dot(x, parameters["weights"]) + parameters["bias"]
+    hypothesis_x = jax.nn.softmax(z,axis=1)
+    regularizer = (lambd * jnp.sum(jnp.sum(parameters["weights"]**2,axis=0)))/(2*x.shape[0])
+    
+    return (-jnp.sum(y*jnp.log(hypothesis_x + 1e-7))/x.shape[0] + regularizer)
+
+def init_weights(n_f, n_c, random_key):
+    """
+    Arguments:
+        n_f : no. of features
+        n_c : no. of classes
+        random_key = unique key to generate pseudo random numbers
+    return:
+        parameters: a dictionary contains weights and bias    
+    """
+    parameters = {}
+    parameters["weights"] = jax.random.normal(key = jax.random.PRNGKey(random_key), shape = [n_f,n_c])
+    parameters["bias"] = jnp.zeros((1,n_c))
+    return parameters
+    
+def fit(x, y, max_iter = 1000, learning_rate = 0.75, lambd = 0.001, random_key = 1):
+    """
+    Arguments:
+        x = training dataset, shape = (no. of examples, no. of features)
+        y = targets, shape = (no. of example, )
+        max_iter = maximum no. of iteration algorithms run
+        learning_rate = stepsize to update the parameters
```

Comment:
The comment string should say that you use adam (otherwise you need to read the code to figure this out).
I suggest also making a BFGS version.

Reply:
I have added a comment for adam optimizer and also implemented the BFGS method but I have to generalize the algorithm for the multi-class problems.